src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs

```diff
@@ -2132,5 +2132,38 @@ public bool IsThrowStatement(SyntaxNode node)
 
         public bool IsLocalFunction(SyntaxNode node)
             => node.IsKind(SyntaxKind.LocalFunctionStatement);
+
+        public SyntaxNode GetIOperationRootNode(SyntaxNode syntaxNode)
+        {
+            Contract.ThrowIfNull(syntaxNode);
+
+            var rootNode = syntaxNode
+                .AncestorsAndSelf(ascendOutOfTrivia: false)
+                .Where(IsRootOperationNode)
+                .LastOrDefault();
```

Comment:
is LastOrDefault appropriate for thigns like nested lambdas?  are you actually wanting to get the topmost node in that case?

Reply:
this needs explanations in code for maintainability.

Reply:
Getting rid of this entirely and moving to something that uses the operation tree directly.